projects/transformers/trainer_mixins/rigl.py

```diff
@@ -0,0 +1,168 @@
+                         steps before pruning for the first time
+    """
+
+    def __init__(self, *args, **kwargs):
+
+        super().__init__(*args, **kwargs)
+        mixin_args = self.args.trainer_mixin_args
+
+        self.prune_fraction = mixin_args.get("prune_fraction", 0.3)
+        self.prune_freq = mixin_args.get("prune_freq", 100)
+        self.warmup_steps = mixin_args.get("warmup_steps", self.prune_freq)
+        self.prune_scheduler = CosineDecayPruneScheduler(
+            total_steps=self.args.max_steps,
+            prune_fraction=self.prune_fraction,
+            warmup_steps=self.warmup_steps
+        )
+        self.sparse_modules = None
+
+    def training_step(self, model, inputs):
+        """Prune and regrow weights every 'prune_freq' iterations."""
+
+        train_loss = super().training_step(model, inputs)
+
+        if self.state.global_step % self.prune_freq != 0:
+            self.prune_scheduler.step()
+            return train_loss
+
+        # Retrieve sparse modules (e.g. SparseWeights) after model has been setup for
+        # distributed training, if it has.
+        if self.sparse_modules is None:
+            self.sparse_modules = filter_modules(
+                model, include_modules=[SparseWeightsBase]
+            ).values()
+
+        # Pre-prune sparsities.
+        param_sparsity0, mask_sparsity0 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Prune weights.
+        model.apply(rezero_weights)
+        prune_fraction = self.prune_scheduler.get_prune_fraction()
+        num_removed = global_prune_by_abs_weight(self.sparse_modules, prune_fraction)
+        model.apply(rezero_weights)
+
+        # Post-prune sparsities.
+        param_sparsity1, mask_sparsity1 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Accumulate gradients over one batch.
+        self.optimizer.zero_grad()
+        train_dataloader = self.callback_handler.train_dataloader
+        train_batch = next(iter(train_dataloader))
+        inputs_to_device(train_batch, device=self.args.device)
+        batch_loss = self.compute_loss(model, train_batch)
+        batch_loss.backward()
+
+        # Regrow weights
+        num_add = self.prune_scheduler.get_num_add(num_removed)
+        global_add_by_abs_grad(self.sparse_modules, num_add)
+        self.prune_scheduler.step()
+
+        # Post-grow sparsities.
+        param_sparsity2, mask_sparsity2 = calc_cumulative_sparsity(self.sparse_modules)
+
+        # Log pruning stats.
+        actual_pruned = param_sparsity1 - param_sparsity0
+        actual_pruned_on_params = actual_pruned / (1 - mask_sparsity0)
+
+        logging.info(f"RigLMixin:")
+        logging.info(f"Target: remove {prune_fraction} frac of on params")
+        logging.info(f"Actual: removed {actual_pruned_on_params} fraction of on params")
+
+        # For now, the logs are very robust to ensure pruning occurs as expected.
+        # TODO: Remove non-essential logging.
```

Comment:
I was gonna comment on putting all these logging under a verbose flag, but I see there is a todo. If that can cut down any performance or memory overhead, it will be useful in the larger runs (although I understand why the logging is relevant for experimentation)

Reply:
I'll adjust these logs in a soon-to-come PR